crates/client/settlement/src/sync_state.rs

```diff
@@ -0,0 +1,207 @@
+                        true => {
+                            log::info!("[settlement] Retrying after error");
+                            continue;
+                        }
+                        false => {
+                            log::error!("[settlement] Unrecoverable error, exiting");
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn sync_state_proc<SP>(substrate_client: &SC, settlement_provider: &SP) -> anyhow::Result<()>
+    where
+        SP: ?Sized + SettlementProvider,
+    {
+        let mut last_settled_state = Self::get_starknet_state(settlement_provider).await?;
+        log::debug!("[settlement] Initial state {:?}", last_settled_state);
+
+        let last_settled_level: u64 = last_settled_state.block_number.try_into()?;
+        log::info!("[settlement] Last settled level {}", last_settled_level);
+
+        let mut finality_notifications = substrate_client.finality_notification_stream();
+
+        while let Some(notification) = finality_notifications.next().await {
+            let block = find_starknet_block(notification.header.digest())
+                .map_err(|e| anyhow::anyhow!("Failed to find SN block: {}", e))?;
+
+            let mut sync_from = last_settled_level;
+            let sync_to = block.header().block_number;
+
+            if sync_from > sync_to {
+                log::info!("[settlement] Skipping block {} (already settled)", sync_to);
+                continue;
+            }
+
+            if sync_from == sync_to {
+                log::info!("[settlement] Verifying state root for block {}", sync_to);
+                Self::verify_starknet_state(substrate_client, &last_settled_state)?;
+                continue;
+            }
+
+            log::info!("[settlement] New finalized block, syncing {} -> {}", sync_from, sync_to);
+            while sync_from < sync_to {
+                let (next_block, substrate_block_hash) = if sync_from + 1 == sync_to {
+                    // This is a typical scenario when we are up to speed with the chain
+                    (block.clone(), notification.hash)
+                } else {
+                    Self::get_starknet_block(substrate_client, sync_from + 1)?
+                };
+
+                let new_state = Self::update_starknet_state(
+                    substrate_client,
+                    settlement_provider,
+                    &last_settled_state,
+                    &next_block,
+                    substrate_block_hash,
+                )
+                .await?;
+
+                log::debug!("[settlement] State transitioned to {:?}", new_state);
+                last_settled_state = new_state;
+                sync_from += 1;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_starknet_block(substrate_client: &SC, block_number: u64) -> anyhow::Result<(StarknetBlock, B::Hash)> {
```

Comment:
Can we move this in some sort of client/utils crate?

Reply:
There is probably code to share with the RPCs

Reply:
I tried to reuse as much as possible, also here we rely on crate-level error type (now)